lib/Frontend/CompilerInvocation.cpp

```diff
@@ -1210,6 +1210,40 @@ static bool ParseTBDGenArgs(TBDGenOptions &Opts, ArgList &Args,
   return false;
 }
 
+/// Determine which frame pointer setting to use. Per default if no flag is set
+/// we will use clang's default setting. This case is denoted by returning None.
+static Optional<clang::CodeGenOptions::FramePointerKind>
+determineFramePointerUse(ArgList &Args) {
+  bool requiresLeafFramePointers =
+      Args.hasArg(options::OPT_no_omit_frame_pointer_leaf);
+  bool omitAllFP = Args.hasArg(options::OPT_omit_frame_pointer);
+  bool omitLeafFP = Args.hasArg(options::OPT_omit_frame_pointer_leaf);
+  bool requiresAllFP = Args.hasArg(options::OPT_no_omit_frame_pointer);
+
+  // If the user explicitly requests -omit-frame-pointer do we emit no
+  // frame pointers.
+  if (omitAllFP)
+    return clang::CodeGenOptions::FramePointerKind::None;
+
+  // If the user explicitly requests -omit-frame-pointer do we emit no
+  // frame pointers.
+  if (omitLeafFP)
+    return clang::CodeGenOptions::FramePointerKind::NonLeaf;
+
+  // Otherwise, if the users requires leaf frame pointer we emit all frame
+  // pointers.
+  if (requiresLeafFramePointers)
+    return clang::CodeGenOptions::FramePointerKind::All;
```

Comment:
Assumes a default of no_omit_frame_pointer. I don't know if that's intended.
What if we also have requiresFP?

Reply:
This is intended.

In the branch we can only have:
-fno-omit-frame-pointer-leaf in which case frame-pointer="all" makes sense
-fno-omit-frame-pointer-leaf -fno-omit-frame-pointer in which case "all" also makes sense.